modules/widgets/migrate-to-core/gallery-widget.php

```diff
@@ -0,0 +1,192 @@
+<?php
+/**
+ * Migration from Jetpack's Gallery Widget to WordPress' Core Gallery Widget.
+ *
+ * @since 5.5
+ *
+ * @package Jetpack
+ */
+/**
+ * Migrates all active instances of Jetpack's Gallery widget to Core's Media Gallery widget.
+ */
+function jetpack_migrate_gallery_widget() {
+	// Only trigger the migration from wp-admin and outside unit tests
+	if ( ! is_admin() || defined( 'PHPUNIT_JETPACK_TESTSUITE' ) ) {
+		return;
+	}
+
+	// Only migrate if the new widget is available and we haven't yet migrated
+	if ( ! class_exists( 'WP_Widget_Media_Gallery' ) || Jetpack_Options::get_option( 'gallery_widget_migration' ) ) {
+		return;
+	}
+
+	$old_widgets = get_option( 'widget_gallery', array() );
+	$media_gallery = get_option( 'widget_media_gallery', array() );
+	$sidebars_widgets = wp_get_sidebars_widgets();
+
+	// Array to store legacy widget ids in to unregister on success.
+	$widgets_to_unregister = array();
+
+	$old_widgets = array_filter( $old_widgets, 'jetpack_migrate_gallery_widget_is_importable' );
+	foreach ( $old_widgets as $id => $widget ) {
+		$new_id = $id;
+		// Try to get an unique id for the new type of widget.
+		// It may be the case that the user has already created a core Gallery Widget
+		// before the migration begins. (Maybe Jetpack was deactivated during core's upgrade).
+		for( $i = 0; $i < 10 && in_array( $new_id, array_keys( $media_gallery ) ); $i++, $new_id++ );
+
+		$widget_copy = jetpack_migrate_gallery_widget_upgrade_widget( $widget );
+
+		if ( null === $widget_copy ) {
+			jetpack_migrate_gallery_widget_bump_stats( 'gallery-widget-skipped' );
+			continue;
+		}
+
+		$media_gallery[ $new_id ] = $widget_copy;
+
+		$sidebars_widgets = jetpack_migrate_gallery_widget_update_sidebars( $sidebars_widgets, $id, $new_id );
+
+		$widgets_to_unregister[ $id ] = $new_id;
+	}
+
+	if ( update_option( 'widget_media_gallery', $media_gallery ) ) {
+		delete_option( 'widget_gallery' );
+
+		// Now un-register old widgets and register new.
+		foreach ( $widgets_to_unregister as $id => $new_id ) {
+			wp_unregister_sidebar_widget( "gallery-${id}" );
+
+			// register new widget.
+			$media_gallery_widget = new WP_Widget_Media_Gallery();
+			$media_gallery_widget->_set( $new_id );
+			$media_gallery_widget->_register_one( $new_id );
+		}
+
+		wp_set_sidebars_widgets( $sidebars_widgets );
+
+		// Log if we migrated all, or some for this site.
+		foreach ( $widgets_to_unregister as $w ) {
+			jetpack_migrate_gallery_widget_bump_stats( 'gallery-widget-migrated' );
+		}
+
+		// We need to refresh on widgets page for changes to take effect.
+		// The jetpack_refresh_on_widget_page function is already defined in migrate-to-core/image-widget.php
+		add_action( 'current_screen', 'jetpack_refresh_on_widget_page' );
+	} else {
+		$widget_media_gallery = get_option( 'widget_media_gallery' );
+		if ( is_array( $widget_media_gallery ) ) {
```

Comment:
what is the expected case we're protecting against here? Should we really be deleting the widget_gallery?

Reply:
This was a sort of cleanup for the case when this script was run more than once. If `widget_media_gallery` didn't need to change when updated, it's because we had already migrated and then it was safe to delete the old ones (But then `update_options` can return false too if it failed, so the rationale is buggy indeed). 

Also, this `else` may no longer be useful after implementing the logic of having unique ids for each newly created `media-gallery` which is only useful if there were any Core Media Gallery widgets before this migration is run. Is this a valid consideration? 

So, yes, deleting is not necessary. Specially if we're to backup the original widgets.

Reply:
Updated in 02bc7df. Now we don't remove the original `widget_gallery` option. Would this work as a backup of the original widgets in case something goes wrong ?